libcontainer/system/proc_test.go

```diff
@@ -1,45 +1,95 @@
+		PID:       9534,
+		Name:      "cat",
+		State:     'R',
+		StartTime: 9214966,
+	},
+	"12345 ((ugly )pr()cess() R 9323 9534 9323 34828 9534 4194304 95 0 0 0 0 0 0 0 20 0 1 0 9214966 7626752 168 18446744073709551615 4194304 4240332 140732237651568 140732237650920 140570710391216 0 0 0 0 0 0 0 17 1 0 0 0 0 0 6340112 6341364 21553152 140732237653865 140732237653885 140732237653885 140732237656047 0": {
+		PID:       12345,
+		Name:      "(ugly )pr()cess(",
+		State:     'R',
+		StartTime: 9214966,
+	},
+	"24767 (irq/44-mei_me) S 2 0 0 0 -1 2129984 0 0 0 0 0 0 0 0 -51 0 1 0 8722075 0 0 18446744073709551615 0 0 0 0 0 0 0 2147483647 0 0 0 0 17 1 50 1 0 0 0 0 0 0 0 0 0 0 0": {
+		PID:       24767,
+		Name:      "irq/44-mei_me",
+		State:     'S',
+		StartTime: 8722075,
+	},
+}
 
-		"24767 (irq/44-mei_me) S 2 0 0 0 -1 2129984 0 0 0 0 0 0 0 0 -51 0 1 0 8722075 0 0 18446744073709551615 0 0 0 0 0 0 0 2147483647 0 0 0 0 17 1 50 1 0 0 0 0 0 0 0 0 0 0 0": {
-			PID:       24767,
-			Name:      "irq/44-mei_me",
-			State:     'S',
-			StartTime: 8722075,
-		},
-	}
-	for line, expected := range data {
+func TestParseStat(t *testing.T) {
+	for line, exp := range procdata {
 		st, err := parseStat(line)
 		if err != nil {
-			t.Fatal(err)
-		}
-		if st.PID != expected.PID {
-			t.Fatalf("expected PID %q but received %q", expected.PID, st.PID)
-		}
-		if st.State != expected.State {
-			t.Fatalf("expected state %q but received %q", expected.State, st.State)
+			t.Errorf("input %q, unexpected error %v", line, err)
+		} else if !reflect.DeepEqual(st, exp) {
+			t.Errorf("input %q, expected %+v, got %+v", line, exp, st)
 		}
-		if st.Name != expected.Name {
-			t.Fatalf("expected name %q but received %q", expected.Name, st.Name)
+	}
+}
+
+func TestParseStatBadInput(t *testing.T) {
+	cases := []struct {
+		desc, input string
+	}{
+		{"no (",
+			"123 ) S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0"},
+		{"no )",
+			"123 ( S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0"},
+		{"negative pid",
+			"-1 (cmd) S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0"},
+		{"empty line",
+			""},
+		{"short line",
+			"123 (cmd) S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0"},
+		{"short line (no space after stime)",
+			"123 (cmd) S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 42"},
+		{"bad stime",
+			"123 (cmd) S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 -1 "},
+		{"bad stime 2", // would be valid if not -1
+			"123 (cmd) S                   -1 "},
+	}
+	for _, c := range cases {
+		st, err := parseStat(c.input)
+		t.Logf("case %q: error: %v, stat: %v", c.desc, err, st)
+		if err == nil {
+			t.Errorf("case %q, expected error, got nil", c.desc)
 		}
```

Comment:
Given that each test has a description, could use sub-tests here;

```suggestion
	for _, c := range cases {
		c := c
		t.Run(c.desc, func(t *testing.T) {
			st, err := parseStat(c.input)
			t.Logf("error: %v, stat: %v", err, st)
			if err == nil {
				t.Fatalf("expected error, got nil", c.desc)
			}
		})
```

Reply:
I'm not sure it makes much sense here, for such trivial checks... the only differences I can think of are more lines of test output, and (apparently) more code (which IMHO decreases the readability). Or am I missing something?

Reply:
I personally prefer `t.Run` because it makes it possible to run a single subtest and it better represents the idea of having the same test code for many cases. But yeah this test is simple enough that `t.Run` probably adds too little to be worth it.